### PR TITLE
Adding library_owners.json to the CLI

### PR DESCRIPTION
**Pull request by andreaangiolillo**
**Issue:**
The Kub team recently added new dependencies to AtlasCLI and agreed to maintain them. This means that they will be responsible for keeping those libraries updated. As more teams will contribute to the CLI, we need to track which teams own which dependencies as the APIx should not be the only one responsible for upgrading the CLI libraries.

**Proposal:**
This is a proposal to add a `library_owners.json` file that contains all the CLI libraries and related owners as we are currently doing in other repos. 

**Validations:** 
- Added check on the pre-commit hook
- Added evergreen task `library_owners_check` to the `code_health`
- Added git action

**Next Steps:**
- Update [MongoDB Atlas CLI (Atlas CLI) and MongoDB CLI (mongocli)](https://github.com/mongodb/mongodb-atlas-cli/pull/1499/files#diff-33ef32bf6c23acb95f5902d7097b7a1d5128ca061167ec0716715b0b9eeaa5f6:~:text=MongoDB%20Atlas%20CLI%20(Atlas%20CLI)%20and%20MongoDB%20CLI%20(mongocli)%20to%20highlight%20that%20we%20will%20tag%20your%20team%20to%20review%20the%20dependabot%20PRs%20related%20to%20these%20dependencies.) to highlight that the team owing the library will review the dependabot PR.
- Create a ticket to investigate if we can add the correct reviewer in the dependabot PR using `library_owners.json` (https://jira.mongodb.org/browse/CLOUDP-149562)

## Checklist

- [x] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added any necessary documentation in document requirements section listed in [CONTRIBUTING.md](https://github.com/mongodb/mongodb-atlas-cli/blob/master/CONTRIBUTING.md) (if appropriate)
- [x] I have addressed the @mongodb/docs-cloud-team comments (if appropriate)
- [x] I have updated [test/README.md](https://github.com/mongodb/mongodb-atlas-cli/blob/master/test/README.md) (if an e2e test has been added)
- [x] I have run `make fmt` and formatted my code

## Further comments

**Example of failed github action job:**
https://github.com/mongodb/mongodb-atlas-cli/actions/runs/3630407651/jobs/6123717699

**Example of a failed run with the MakeFile**
```
make check-library-owners                                                                                                                                               1.19
==> Check library_owners.json
go run ./tools/libraryowners/main.go
2022/12/06 14:00:26
'github.com/google/go-github/v42' is not inside 'library_owners.json'. Please, add this dependency to 'library_owners.json'.
exit status 1
make: *** [check-library-owners] Error 1
```

### CODE REVIEW

**Review by blva — DISMISSED** [review]

**Review by gssbzn — CHANGES_REQUESTED** [review]
not worth adding without a test to enforce it

File: tools/libraryowners/main.go [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+import (
+	"encoding/json"
+	"log"
+	"os"
+
+	"golang.org/x/mod/modfile"
+)
+
+const libraryOwnersPath = "library_owners.json"
+const goModpath = "go.mod"
```
Comment (gssbzn) [inline tools/libraryowners/main.go:26]: [nit] let's group them 
```suggestion
const (
libraryOwnersPath = "library_owners.json"
goModpath = "go.mod"
)
```

File: tools/libraryowners/main.go [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+	libraryOwnersFile, err := os.ReadFile(libraryOwnersPath)
+	if err != nil {
+		log.Fatalln(err)
+	}
+
+	if len(libraryOwnersFile) == 0 {
+		log.Fatalf("\n'%s' is empty", libraryOwnersPath)
+	}
+
+	var libraryOwnersContent map[string]interface{}
```
Comment (gssbzn) [inline tools/libraryowners/main.go:51]: [q] wouldn't this be
```suggestion
	var libraryOwnersContent map[string]string
```

File: tools/libraryowners/main.go [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+}
+
+func validate(libraryOwner map[string]interface{}, goMod *modfile.File) {
+	for _, library := range goMod.Require {
+		if library.Indirect {
+			continue
+		}
+
+		if val, ok := libraryOwner[library.Mod.Path]; !ok {
+			log.Fatalf("\n'%s' is not inside '%s'. Please, add this dependency to '%s'.", library.Mod.Path, libraryOwnersPath, libraryOwnersPath)
```
Comment (gssbzn) [inline tools/libraryowners/main.go:66]: this is probably the other way around, remove the ownership kind of thing, you may also be able to automate the removal of the ownership but the error is fine for now

File: .github/pull_request_template.md [hunk 4]
```diff
@@ -35,6 +35,7 @@ We're here to help! This is simply a reminder of what we are going to look for b
 - [ ] I have added any necessary documentation in document requirements section listed in [CONTRIBUTING.md](https://github.com/mongodb/mongodb-atlas-cli/blob/master/CONTRIBUTING.md) (if appropriate)
 - [ ] I have addressed the @mongodb/docs-cloud-team comments (if appropriate)
 - [ ] I have updated [test/README.md](https://github.com/mongodb/mongodb-atlas-cli/blob/master/test/README.md) (if an e2e test has been added)
+- [ ] I have updated [library_owners.json](https://github.com/mongodb/mongodb-atlas-cli/blob/master/library_owners.json) (if a new library has been added)
```
[the remark this answers is not in the document]
Comment (andreaangiolillo) [reply .github/pull_request_template.md]: removed it

File: .github/workflows/library-owners.yml [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+          go-version: 1.19
+      - name: Cache Dependencies
+        uses: actions/cache@v3
+        with:
+          path: ~/go/pkg/mod
+          key: Linux-go-1.19-${{ hashFiles('**/go.sum') }}
+          restore-keys: |
+            Linux-go-1.19-
+      - name: Intsall go-mock
+        run: go install github.com/golang/mock/mockgen@latest
```
[the remark this answers is not in the document]
Comment (andreaangiolillo) [reply .github/workflows/library-owners.yml]: ops, leftover. I removed it

File: .github/workflows/library-owners.yml [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+      - name: Run check-library-owners
+        run: make check-library-owners
```
[the remark this answers is not in the document]
Comment (andreaangiolillo) [reply .github/workflows/library-owners.yml]: updated

File: library_owners.json [hunk 7]
```diff
@@ -0,0 +1,43 @@
+{
```
[the remark this answers is not in the document]
Comment (andreaangiolillo) [reply library_owners.json]: moved to `build/ci`

File: tools/libraryowners/main.go (hunk 1 above)
Comment (andreaangiolillo) [reply tools/libraryowners/main.go]: updated

File: tools/libraryowners/main.go (hunk 2 above)
Comment (andreaangiolillo) [reply tools/libraryowners/main.go]: updated

File: tools/libraryowners/main.go (hunk 3 above)
Comment (andreaangiolillo) [reply tools/libraryowners/main.go]: updated

File: tools/libraryowners/main.go [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+	}
+
+	if len(libraryOwnersFile) == 0 {
+		log.Fatalf("\n'%s' is empty", libraryOwnersPath)
+	}
+
+	var libraryOwnersContent map[string]interface{}
+	err = json.Unmarshal(libraryOwnersFile, &libraryOwnersContent)
+	if err != nil {
+		log.Fatal("\nError during Unmarshal(): ", err)
```
[the remark this answers is not in the document]
Comment (andreaangiolillo) [reply tools/libraryowners/main.go]: updated

File: tools/libraryowners/main.go [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+			return fmt.Errorf("'%s' doesn't have a owner. Please, add a owner to %s in '%s", library.Mod.Path, library.Mod.Path, libraryOwnersPath)
+		}
+	}
+
+	return nil
+}
+
+func main() {
+	if err := validate(newLibraryOwners(), newGoMod()); err != nil {
+		fmt.Printf("Error during the validation of '%s': %v\n", libraryOwnersPath, err)
```
Comment (gssbzn) [inline tools/libraryowners/main.go:84]: [nit] let's print to stderr

File: tools/libraryowners/main.go [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+	if len(libraryOwnersFile) == 0 {
+		fmt.Printf("'%s' is empty: %v\n", libraryOwnersPath, err)
+		os.Exit(1)
+	}
+
+	var libraryOwnersContent map[string]string
+	err = json.Unmarshal(libraryOwnersFile, &libraryOwnersContent)
+	if err != nil {
+		fmt.Printf("Error during Unmarshal(): %v\n", err)
+		os.Exit(1)
```
Comment (gssbzn) [inline tools/libraryowners/main.go:61]: [nit] this is no different from log fatal ;)

File: tools/libraryowners/main.go [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+	goModFile, err := os.ReadFile(goModpath)
+	if err != nil {
+		fmt.Printf("error reading %s: %v\n", goModpath, err)
+		os.Exit(1)
+	}
+
+	f, err := modfile.Parse(goModpath, goModFile, nil)
+	if err != nil {
+		fmt.Printf("error parsing %s: %v\n", goModpath, err)
+		os.Exit(1)
```
Comment (gssbzn) [inline tools/libraryowners/main.go:40]: fmt.Printf and then os.Exit(1) is no different from log.Fatal, by best practices I mean returning an erro and have the erro be handled 

IE: [Don't panic](https://github.com/golang/go/wiki/CodeReviewComments#dont-panic) 
Where log.Fatal or this approach is no sodifferent from a panic

**Review by gssbzn — COMMENTED** [review]
General LGTM but I think colm's comments need to be address and there's an internal conversation about it

**Comment by andreaangiolillo** [comment]
ah,l I missed the internal conversation. Thanks!

File: library_owners.json (hunk 7 above)
[the remark this answers is not in the document]
Comment (andreaangiolillo) [reply library_owners.json]: updated the file to use GitHub Team names

File: tools/libraryowners/main.go (hunk 9 above)
Comment (andreaangiolillo) [reply tools/libraryowners/main.go]: updated

File: tools/libraryowners/main.go (hunk 10 above)
Comment (andreaangiolillo) [reply tools/libraryowners/main.go]: updated the function to return err instead. Thanks!

**Review by gssbzn — APPROVED** [review]
LGTM thanks for all the work here

**Review by igor-karpukhin — APPROVED** [review]

**Pull request merged**